### DISCUSSION TITLE
How to link a serial port to Wineskinserver?

### DISCUSSION THREAD

**Discussion by RobNelson414** [The-Wineskin-Project/WineskinServer#267] [Q&A] [answered]
Hello,
I have a Windows program that flashes firmware into hardware.  This uses a serial comport such as COM1. It normally runs on Windows.  I want to run it on a Mac Mini using Big Sur MacOS.
I successfully created a wineskin and can open the program on the Mac. 
I need to know how to tell Wineskin server that the USB device named AK0703SA (as shown in Mac system report on BigSur OS) is COM1. some articles say this is automatic.  some say you need to do a symbiotic link, or a regedit (or both). 
 
i've read a few WINE articles that talk about the USB devices being listed in a folder viewable through terminal at ls -l ~/.wine/dosdevices/ 
I don't seem to have that folder. 
I made a wineskin server on one machine then copied it to another so its ready to go (like a virtual machine would be).  Do i have to install it or something in order to get the /.wine/dosdevices/ to be created or something?

**Comment by emendelson** [comment, marked as answer]
In Wineskin, dosdevices is a folder inside the wrapper. The exact location differs depending on when you created the wrapper, but it's easy to find. In a newly-created wrapper it's likely to be in Contents/SharedSupport/prefix

**Comment by 74anders** [comment]
Hello! i had the same issue... but i created the com1 link to  the right tty adress... But still problem that the software says no connection with com1.. Any idea?

**Reply by NormMan** [reply]
Wineskin issue with USB Serial Adaptors on some Apps under El Capitan

This might solve your problem, might not, but it can serve as a help for anyone else who drops in here with a serial port issue.

I've used wineskin on my Macs since my Windows Industial Stength Laptop failed me.

I was able to get my favourite applications to run with only a few minor niggles.

That was on Snow Leopard.

With El Capitan I really struggled to get some of my builds to work even though they had worked fine on Snow Leopard.

Attempts to build wrappers with Winery 1.8.4.2 always failed with WS9Wine2.22 is corrupt.

I then discovered that I could build wrappers with Wineskin 2.6.2 and WS9Wine2.22 if I used Winery 1.8.4.1

That left me with the niggle that I had to enable the native X11 instead, but it worked.

I then found my USB-Serial adaptor was a problem. Some applications would see it and use it with no issue, some would not. I tried another adaptor but that has the same issue. Some Apps would see it, some wouldn't.

I had already added symlinks under dosdevices and that had enabled some to work, but again not others.

After a lot of digging I seemed that something was happening during enumeration that removed and failed to reinstall the serial ports every time wine started the application. El Capitan is very fussy about privilege when trying access to /dev

I tried a lot but eventually discovered that I could force the situation by starting from a batch file that had previously created the appropriate registry entries.

E.g.
the batch script StartApp.bat contains

Regedit Comports.reg
start "" MyFavouriteApp.exe
exit

ComPorts.reg contains
Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\HARDWARE\DEVICEMAP\SERIALCOMM]
"Serial0"="COM1"
"Serial1"="COM2"
"Serial2"="COM3"

These relate to the three serial devices registed in HKEY_LOCAL_MACHINE\System\CurrentControlSet\Enum\SERIAL\COMx

E.g.[HKEY_LOCAL_MACHINE\System\CurrentControlSet\Enum\SERIAL\COM1\COM1]
"ClassGUID"="4D36E978-E325-11CE-BFC1-08002BE10318"
"FriendlyName"="USB Serial Port (COM1)"
"PortName"="COM1"
and so on for the other two adaptors

The batch file will register the necessary serial ports, then start the Application and exit the shell.

This is the way I have made all the awkard apps run on El Capitan.

**Reply by Gcenx** [reply]
> Wineskin issue with USB Serial Adaptors on some Apps under El Capitan

This isn’t a Wineskin issue but a wine issue.

> This might solve your problem, might not, but it can serve as a help for anyone else who drops in here with a serial port issue.
> 
> I've used wineskin on my Macs since my Windows Industial Stength Laptop failed me.
> 
> I was able to get my favourite applications to run with only a few minor niggles.
> 
> That was on Snow Leopard.

That reminds me I had worked with a macports maintainer on getting a more modern version of wine working on 10.6 though I doubt anyone would be willing to spend the time needed to compile it.

> With El Capitan I really struggled to get some of my builds to work even though they had worked fine on Snow Leopard.

That’s not too surprising as I’ve long dropped support for legacy versions of macOS, the bundled copy of 7z might not even work below 10.13

> Attempts to build wrappers with Winery 1.8.4.2 always failed with WS9Wine2.22 is corrupt.

These ancient engines have a hard requirement of XQuartz-2.7.7

> I then discovered that I could build wrappers with Wineskin 2.6.2 and WS9Wine2.22 if I used Winery 1.8.4.1

If you happen to have a clean copy of Wineskin-2.6.2 it might be hardly to uploaded them somewhere for others stuck on legacy versions of macOS.

> That left me with the niggle that I had to enable the native X11 instead, but it worked.

The usb support isn’t tied to using winemac.drv or winex11.drv. 

> I then found my USB-Serial adaptor was a problem. Some applications would see it and use it with no issue, some would not. I tried another adaptor but that has the same issue. Some Apps would see it, some wouldn't.
> 
> I had already added symlinks under dosdevices and that had enabled some to work, but again not others.
> 
> After a lot of digging I seemed that something was happening during enumeration that removed and failed to reinstall the serial ports every time wine started the application. El Capitan is very fussy about privilege when trying access to /dev

macOS El Capitan is when SIP was introduced that would explain the issues you’ve had touching /dev as that’s considered a system owned location.

> I tried a lot but eventually discovered that I could force the situation by starting from a batch file that had previously created the appropriate registry entries.
> 
> E.g. the batch script StartApp.bat contains
> 
> Regedit Comports.reg start "" MyFavouriteApp.exe exit
> 
> ComPorts.reg contains Windows Registry Editor Version 5.00
> 
> [HKEY_LOCAL_MACHINE\HARDWARE\DEVICEMAP\SERIALCOMM] "Serial0"="COM1" "Serial1"="COM2" "Serial2"="COM3"
> 
> These relate to the three serial devices registed in HKEY_LOCAL_MACHINE\System\CurrentControlSet\Enum\SERIAL\COMx
> 
> E.g.[HKEY_LOCAL_MACHINE\System\CurrentControlSet\Enum\SERIAL\COM1\COM1] "ClassGUID"="4D36E978-E325-11CE-BFC1-08002BE10318" "FriendlyName"="USB Serial Port (COM1)" "PortName"="COM1" and so on for the other two adaptors
> 
> The batch file will register the necessary serial ports, then start the Application and exit the shell.

While I’ll lightly never personally need this I’m sure other users may find this useful so thank you.

> This is the way I have made all the awkard apps run on El Capitan.

Assuming your referring to applications that your using that require serial port access.
